alf/algorithms/mi_estimator.py

```diff
@@ -0,0 +1,124 @@
+
+
+class MINEstimator(Algorithm):
+    """Mutual Infomation Neural Estimator.
+
+    Implements MINE and MINE-f estimator in paper
+    Belghazi et al "Mutual Information Neural Estimation"
+    http://proceedings.mlr.press/v80/belghazi18a/belghazi18a.pdf
+    """
+
+    def __init__(self,
+                 x_spec: tf.TensorSpec,
+                 y_spec: tf.TensorSpec,
+                 model=None,
+                 fc_layers=(256, ),
+                 buffer_size=65536,
+                 optimizer: tf.optimizers.Optimizer = None,
+                 estimator_type='MINE',
+                 averager=ScalarAdaptiveAverager(),
+                 name="MIEstimator"):
+        """Create a MIEstimator.
+
+        Args:
+            x_spec (TensorSpec): spec of x
+            y_spec (TensorSpec): spec of y
+            model (Network): can be called as model([x, y]) and return a Tensor
+                with shape=[batch_size, 1]. If None, a default MLP with
+                fc_layers will be created.
+            fc_layers (tuple[int]): size of hidden layers. Only used if model is
+                None.
+            buffer_size (int): capacity of buffer for storing y
+            optimzer (tf.optimizers.Optimzer): optimizer
+            estimator_type (str): one of 'MINE' and 'MINE-f'
+            averager (EMAverager): averager used to maintain a moving average
+                of exp(T). Only used for 'MINE' estimator
+            name (str): name of this estimator
+        """
+        assert estimator_type in ['MINE', 'MINE-f']
+        super().__init__(train_state_spec=(), optimizer=optimizer, name=name)
+        self._x_spec = x_spec
+        self._y_spec = y_spec
+        if model is None:
+            model = EncodingNetwork(
+                name="MIEstimator",
+                input_tensor_spec=[x_spec, y_spec],
+                fc_layer_params=fc_layers,
+                last_layer_size=1)
+        self._model = model
+        self._type = estimator_type
+        self._y_buffer = DataBuffer(y_spec, capacity=buffer_size)
+        self._z_averager = averager
+
+    def train_step(self, inputs, state=None):
+        """Perform training on one batch of inputs.
+
+        Args:
+            inputs (tuple(Tensor, Tensor)): tuple of x and y
+            state: not used
+        Returns:
+            AlgorithmStep
+                outputs (Tensor): shape=[batch_size], its mean is the estimated
+                    MI
+                state: not used
+                info (LossInfo): info.loss is the loss
+        """
+        (x, y) = inputs
+        num_outer_dims = get_outer_rank(x, self._x_spec)
+        batch_squash = BatchSquash(num_outer_dims)
+        x = batch_squash.flatten(x)
+        y = batch_squash.flatten(y)
+        batch_size = y.shape[0]
+        self._y_buffer.add_batch(y)
+        y1 = self._y_buffer.get_batch(batch_size)
+        log_ratio = self._model([x, y])[0]
+        ratio = tf.math.exp(self._model([x, y1])[0])
```

Comment:
should  check for NaN and Inf values or clip before `exp`

Reply:
added clipping